### PR TITLE
Only register required RESTEasy providers in REST Client

### LINKED ISSUE #707: Only register the required RESTEasy providers in the REST client extension
Right now, the REST client extension registers all the RESTEasy providers, which is a very bad idea.

We should reduce that to the bare minimum based on the `@Consumes/@Produces` annotations. It's probably something we can do by sharing the code already present in the JAX-RS extension.

### PR DESCRIPTION
**Pull request by kenfinnigan**
Fixes #707

Moved provider lookup and discovery into `resteasy-common`.

### CODE REVIEW

File: extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java
Comment (gsmet) [inline extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java]: Maybe it's a stupid question but you don't want to support the gzip providers in the REST client case?

It would probably be nice to support it too, no? I think what is done in the server part indeed also considers the client case so I would just put that code in resteasy-common. And this particular config item would be moved to the common part too.

File: extensions/smallrye-rest-client/runtime/src/main/java/io/quarkus/smallrye/restclient/runtime/RestClientBuilderImpl.java
Comment (gsmet) [inline extensions/smallrye-rest-client/runtime/src/main/java/io/quarkus/smallrye/restclient/runtime/RestClientBuilderImpl.java]: You don't want to support it here?

File: extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java
Comment (gsmet) [inline extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java]: I think the reason why your jar is a bit bigger is that you didn't remove that code: https://github.com/quarkusio/quarkus/pull/1619/files#diff-90fd4c56e0f60bb0cb22626be37bea55R98 . So you add additional reflection rules without removing any code in the end.

Note that you will **need** to support the detected providers in `RestClientBuilderImpl`. Otherwise you will have some issues with providers not being available.

I also think that https://github.com/quarkusio/quarkus/pull/1619/files#diff-90fd4c56e0f60bb0cb22626be37bea55R92 should probably be conditioned to the presence of the JSON binding provider in the list (or the built-in case).

**Review by gsmet — CHANGES_REQUESTED** [review]
Nice cleanup overall but I think I spotted a few issues.

File: extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java
Comment (kenfinnigan) [reply extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java]: Wouldn't the gzip providers be added when the `@Produces` and/or `@Consumes` types indicate gzip?

Do we need those providers if they don't specify gzip?

File: extensions/smallrye-rest-client/runtime/src/main/java/io/quarkus/smallrye/restclient/runtime/RestClientBuilderImpl.java
Comment (kenfinnigan) [reply extensions/smallrye-rest-client/runtime/src/main/java/io/quarkus/smallrye/restclient/runtime/RestClientBuilderImpl.java]: Not sure what you mean.

File: extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java
Comment (kenfinnigan) [reply extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java]: Not sure why but the diff links don't show me anything specific, just this page at the top.

`RestClientBuilderImpl` registers the providers with RESTEasy through the properties. I'm not sure what else you're expecting it to do

File: extensions/smallrye-rest-client/runtime/src/main/java/io/quarkus/smallrye/restclient/runtime/RestClientBuilderImpl.java
Comment (gsmet) [reply extensions/smallrye-rest-client/runtime/src/main/java/io/quarkus/smallrye/restclient/runtime/RestClientBuilderImpl.java]: Well, in the case of `RestClientBuilderImpl`, AFAICS, you use all the built-in providers. Which means that you will need them all to be registered.

File: extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java
Comment (gsmet) [reply extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java]: GitHub is flaky today, I suppose that's why it doesn't bring you to the right line.

Trying with another simpler link: here you still register all the providers for reflection: https://github.com/quarkusio/quarkus/blob/master/extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java#L97 . Which means that they will all be embedded in the native image.

The whole point of studying the `@Consumes`/`@Produces` annotations it to not do that.

Comment (kenfinnigan) [reply extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java]: Ugh, that's my bad, completely missed that!

I imagine the Abstract JSON-B provider a bit further up should also be removed?

File: extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java
Comment (gsmet) [reply extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java]: The gzip feature is totally orthogonal. It's about the client/server supporting gzip HTTP encoding. Typically, your JSON rest service can serve your JSON compressed as gzip and the client automatically deflates it.

That's why it's an option because basically, in the rest client case, you say: "my RESTEasy installation supports gzip, you can send me gzip encoded responses".

File: extensions/smallrye-rest-client/runtime/src/main/java/io/quarkus/smallrye/restclient/runtime/RestClientBuilderImpl.java
Comment (kenfinnigan) [reply extensions/smallrye-rest-client/runtime/src/main/java/io/quarkus/smallrye/restclient/runtime/RestClientBuilderImpl.java]: Setting that property is telling RESTEasy we don't want to use all the built-in providers.

Other than my mess up on the processor, not sure why you see `RestClientBuilderImpl` using all the built-in providers?

Comment (gsmet) [reply extensions/smallrye-rest-client/runtime/src/main/java/io/quarkus/smallrye/restclient/runtime/RestClientBuilderImpl.java]: Ah, yes, it's my mistake. Read it as `public static final` instead of `public static`. This part looks good, sorry for the noise.

**Comment by kenfinnigan** [comment]
No worries @gsmet, working on fixes now.

Having some issues with Intellij at present

File: extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java
Comment (gsmet) [reply extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java]: I think it should be conditioned on the presence of `org.jboss.resteasy.plugins.providers.jsonb.JsonBindingProvider` in the list.

Comment (gsmet) [reply extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java]: But thinking about it more, it should be moved to the common part and be added if `providersToRegister` contains `org.jboss.resteasy.plugins.providers.jsonb.JsonBindingProvider`.

Comment (kenfinnigan) [reply extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java]: Thinking the same, so will do that

**Comment by kenfinnigan** [comment]
@gsmet I think I've sorted everything now

**Review by gsmet — CHANGES_REQUESTED** [review]
@kenfinnigan I added two minor comments. Otherwise, it looks good. Could you take a look?

The bad news is that I will need to upgrade my rest-client guide with the new numbers given the binary size is now reduced.

File: extensions/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java
Comment (kenfinnigan) [reply extensions/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java]: The original Config object was named `ResteasyConfig`, which from the default name settings of `ConfigRoot` I thought meant that would be converted to `resteasy-config`.

If I'm wrong and it's just `resteasy` I can fix

File: extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyScanningProcessor.java
Comment (kenfinnigan) [reply extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyScanningProcessor.java]: Sure, that's a carryover from moving the gzip bits out

File: extensions/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java
Comment (gsmet) [reply extensions/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java]: The `Config` suffix is removed by the config machinery.

Comment (kenfinnigan) [reply extensions/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java]: ah ok, my bad

**Comment by kenfinnigan** [comment]
Done @gsmet

**Review by gsmet — APPROVED** [review]
I rebased and squashed to have one atomic commit.

Looks good to me. Let's wait for CI before merging.

**Comment by kenfinnigan** [comment]
Thanks @gsmet

**Pull request merged**
